linear_programming/simplex.py

```diff
@@ -0,0 +1,327 @@
+"""
+Python implementation of the simplex algorithm for solving linear programs in
+tabular form with
+- `>=`, `<=`, and `=` constraints and
+- each variable `x1, x2, ...>= 0`.
+
+See https://gist.github.com/imengus/f9619a568f7da5bc74eaf20169a24d98 for how
+to convert linear programs to simplex tableaus, and the steps taken in the
+simplex algorithm.
+
+Resources:
+https://en.wikipedia.org/wiki/Linear_programming
+https://en.wikipedia.org/wiki/Simplex_algorithm
+
+https://towardsdatascience.com/ \
+    a-beginners-guide-to-linear-programming-and-the-simplex-algorithm \
+    -87db017e92b4
+"""
+from typing import Any
+
+import numpy as np
+
+
+class Tableau:
+    """Operate on simplex tableaus"""
+
+    def __init__(self, tableau: np.ndarray, n_vars: int) -> None:
+        """
+        >>> tableau = np.array([[-1,-1,0,0,-1],[1,3,1,0,4],[3,1,0,1,4.]])
+        >>> t = Tableau(tableau, 2)
+        Traceback (most recent call last):
+        ...
+        ValueError: RHS must be > 0
+        """
+        rhs = tableau[:, -1]
+        if np.any(rhs, where=rhs < 0):
+            raise ValueError("RHS must be > 0")
+
+        if tableau is None:
+            return
+
+        self.tableau = tableau
+        self.n_rows = len(tableau[:, 0])
+        self.n_cols = len(tableau[0])
+
+        # Number of decision variables x1, x2, x3...
+        self.n_vars = n_vars
+
+        # Number of artificial variables to be minimised
+        self.n_art_vars = len(np.where(tableau[self.n_vars : -1] == -1)[0])
+
+        # 2 if there are >= or == constraints (nonstandard), 1 otherwise (std)
+        self.n_stages = (self.n_art_vars > 0) + 1
+
+        # Number of slack variables added to make inequalities into equalities
+        self.n_slack = self.n_rows - self.n_stages
+
+        # Objectives for each stage
+        self.objectives = ["max"]
+
+        # In two stage simplex, first minimise then maximise
+        if self.n_art_vars:
+            self.objectives.append("min")
+
+        self.col_titles = [""]
+
+        # Index of current pivot row and column
+        self.row_idx = None
+        self.col_idx = None
+
+        # Does objective row only contain (non)-negative values?
+        self.stop_iter = False
+
+    @staticmethod
+    def generate_col_titles(*args: int) -> list[str]:
+        """Generate column titles for tableau of specific dimensions
+
+        >>> Tableau.generate_col_titles(2, 3, 1)
+        ['x1', 'x2', 's1', 's2', 's3', 'a1', 'RHS']
+        """
+        # decision | slack | artificial
+        string_starts = ["x", "s", "a"]
+        titles = []
+        for i in range(3):
+            for j in range(args[i]):
+                titles.append(string_starts[i] + str(j + 1))
+        titles.append("RHS")
+        return titles
```

Comment:
Much more consise
```suggestion
        return [string_starts[i] + str(j + 1) for j in range(args[i]) for i in range(3), "RHS"]
```

Reply:
This list comprehension doesn't work. For readability's sake, I'll stick with the for loops